### PR TITLE
feat(gwyneth): add first simple tests for proposer/prover

### PR DESCRIPTION
**Pull request by adaki2004**
todos:

- [x] some `proposeBlock()` tests (without `ISequencerRegistry` tests so fer)
- [x] `proveBlock()`
- [x] verify
- [x] some obvious failing edge cases

### CODE REVIEW

**Comment by adaki2004** [comment]
@Brechtpd  proposeBlock() works with X amount of iteration but combined with proveBlock() currently it fails. (Can debug later why), but i do have one question.

Can you check file BasedOperator.sol and line 149, and my comment in the code ? I'm not sure in that logic, but i also know the original also not perfect, because it will revert with block already proven, so somehow we need to find a way if a block was proven before or not, right ?

Thanks !

(if you wanna run, just run : `forge test --match-test test_L1_proposeBlock -vvvv`, but currently it has some debug prints)

File: packages/protocol/contracts/L1/BasedOperator.sol [hunk 1]
```diff
@@ -124,16 +136,27 @@ contract BasedOperator is EssentialContract, TaikoErrors {
...
+        );
+
+        console2.log("What is stored:");
+        console2.logBytes32(storedTransition.blockHash);
+
+        console2.log("What we are trying to prove:");
+        console2.logBytes32(proofBatch.transition.blockHash);
+        
+        // Brecht: SO we set the blockHash in proposeBlock().
+        // But we need to prove it too (the same one), so somehow we need to check if this is proven already and IF NOT, then revert with "block already proven", no ? So i set the verifiableAfter in propseBlock to 0, because this is just a "proposed state".
```
Comment (Brechtpd) [inline packages/protocol/contracts/L1/BasedOperator.sol:150]: ah yes, sounds good

File: packages/protocol/contracts/L1/BasedOperator.sol [hunk 2]
```diff
@@ -71,7 +78,7 @@ contract BasedOperator is EssentialContract, TaikoErrors {
     {
         require(msg.value == PROVER_BOND, "Prover bond not expected");
 
-        _block = taiko.proposeBlock(params, txList);
+        _block = TaikoL1(resolve("taiko", false)).proposeBlock(params, txList);
```
Comment (Brechtpd) [inline packages/protocol/contracts/L1/BasedOperator.sol:81]: I don't know if we should use address manager, but if it makes things easier for now good for me.

File: packages/protocol/contracts/L1/TaikoL1.sol [hunk 3]
```diff
@@ -57,7 +58,7 @@ contract TaikoL1 is EssentialContract, TaikoEvents, TaikoErrors {
         payable
         nonReentrant
         whenNotPaused
-        onlyFromNamed("operator")
+        onlyFromNamed("based_operator")
```
Comment (Brechtpd) [inline packages/protocol/contracts/L1/TaikoL1.sol:61]: In theory TaikoL1 should be general for any type of operator, so I think best to keep the general "operator" name here.

File: packages/protocol/contracts/L1/TaikoL1.sol [hunk 4]
```diff
@@ -93,7 +94,8 @@ contract TaikoL1 is EssentialContract, TaikoEvents, TaikoErrors {
         );
 
         TaikoData.Block storage parentBlock = state.blocks[(state.numBlocks - 1)];
-        require(_block.parentMetaHash == parentBlock.metaHash, "invalid parentMetaHash");
+
+        require(_block.parentMetaHash == parentBlock.blockHash, "invalid parentMetaHash");
```
Comment (Brechtpd) [inline packages/protocol/contracts/L1/TaikoL1.sol:98]: Oh `blockHash` is the actual block hash, meta hash is the hash of the block meta data, so different values. I think the version before was correct or any issues with it?

File: packages/protocol/contracts/L1/BasedOperator.sol (hunk 1 above)
Comment (adaki2004) [reply packages/protocol/contracts/L1/BasedOperator.sol:150]: But actually my comment is incorrect, since verifiableAfter is used on multiple places, so i just used the `isProven` flag.

File: packages/protocol/contracts/L1/BasedOperator.sol (hunk 2 above)
Comment (adaki2004) [reply packages/protocol/contracts/L1/BasedOperator.sol:81]: Personall (for tests also) found it easier, and what if it changes... ? Might revisit later ?

**Review by Brechtpd — APPROVED** [review]
